src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs

```diff
@@ -0,0 +1,142 @@
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var enumType = typeInferenceService.InferType(context.SemanticModel, position, objectAsDefault: true, cancellationToken: cancellationToken);
+
+            if (enumType is not { TypeKind: TypeKind.Enum })
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+
+            var hideAdvancedMembers = options.GetOption(RecommendationOptions.HideAdvancedMembers, context.SemanticModel.Language);
+
+            // We'll want to build a list of the actual enum members and all accessible instances of that enum, too
+            var result = enumType.GetMembers().Where(m =>
+                m.Kind == SymbolKind.Field &&
+                ((IFieldSymbol)m).IsConst &&
+                m.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation)).ToImmutableArray();
+
+            return Task.FromResult(result);
+        }
+
+        protected override Task<ImmutableArray<ISymbol>> GetSymbolsAsync(SyntaxContext context, int position,
+            OptionSet options, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = context.GetLanguageService<ISyntaxFactsService>();
+            var node = syntaxFacts.FindTokenOnLeftOfPosition(context.SyntaxTree.GetRoot(cancellationToken), position, includeSkipped: true).Parent;
+            if (syntaxFacts.IsInNonUserCode(context.SyntaxTree, context.Position, cancellationToken) ||
+                syntaxFacts.IsSkippedTokensTrivia(node)) // Note: IsSkippedTokensTrivia implementation is probably buggy
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            if (context.TargetToken.RawKind == syntaxFacts.SyntaxKinds.DotToken)
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var span = new TextSpan(position, 0);
+            var enumType = typeInferenceService.InferType(context.SemanticModel, position, objectAsDefault: true, cancellationToken: cancellationToken);
+
+            if (enumType is not { TypeKind: TypeKind.Enum })
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            var hideAdvancedMembers = options.GetOption(RecommendationOptions.HideAdvancedMembers, context.SemanticModel.Language);
+
+            var otherSymbols = context.SemanticModel.LookupSymbols(position).WhereAsArray(s =>
+                s.MatchesKind(SymbolKind.Field, SymbolKind.Local, SymbolKind.Parameter, SymbolKind.Property) &&
+                s.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation));
+
+            var otherInstances = otherSymbols.WhereAsArray(s => Equals(enumType, s.GetSymbolType()));
```

Comment:
i don't ynderstand what this code is doing.  can you doc why we need this?

Reply:
Me neither. This is a 100% code port from the existing VB provider. If I remove `otherInstances` these tests fail:
```VB
        <Fact, Trait(Traits.Feature, Traits.Features.Completion)>
        <WorkItem(815963, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/815963")>
        Public Async Function TestLocalNoAs() As Task
            Dim markup = <Text><![CDATA[
Enum E
    A
End Enum
 
Class C
    Sub M()
        Const e As E = e$$
    End Sub
End Class
]]></Text>.Value
            Await VerifyItemExistsAsync(markup, "e")
            Await VerifyItemIsAbsentAsync(markup, "e As E")
        End Function
```
```CS
        [Fact]
        [Trait(Traits.Feature, Traits.Features.Completion)]
        public async Task TestLocalNoAs()
        {
            var markup = @"
enum E
{
    A
}

class C
{
    public void M()
    {
        const E e = e$$;
    }
}
";

            await VerifyItemExistsAsync(markup, "e");
            await VerifyItemIsAbsentAsync(markup, "e as E");
        }
```
```
    No expected item was found.
    Existing items: E, E.A
    Expected: True
    Actual:   False
```

I have no idea why `e` is considered useful here. I think `otherInstances ` can be removed.